Implement isFullWebBrowser check on macOS

szewai opened the pull request:
#### ed62cd603fa0941eca6db6fc2ad512ca473690cb
<pre>
Implement isFullWebBrowser check on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=255668">https://bugs.webkit.org/show_bug.cgi?id=255668</a>
rdar://108271078

Reviewed by Per Arne Vollan.

isFullWebBrowser() is available on Cocoa platforms but only returns meaningful value on non-macOS platforms, because it
is implemented by checking an entitlement that is not avaiable on macOS. This patch implements the check on macOS. To
avoid breakage, this patch does not change any behaivor, by ensuring the existing callsites get the same value as before
this patch.

Also, `isFullWebBrowser()` actually returns true when app is browser *or* test app, so this patch renames it to
`isFullWebBrowserOrRunningTest()`.

* Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.h:
* Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm:
(WebKit::determineTrackingPreventionStateInternal):
(WebKit::isFullWebBrowserOrRunningTest):
(WebKit::isFullWebBrowser): Deleted.
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.cpp:
(API::HTTPCookieStore::filterAppBoundCookies):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadData):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::setIsNavigatingToAppBoundDomainAndCheckIfPermitted):

Canonical link: <a href="https://commits.webkit.org/263198@main">https://commits.webkit.org/263198@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/425c08b87847690767bdc6923ab4b6a102333a67

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3800 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3886 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3994 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5230 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4063 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3964 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3891 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3517 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3845 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4051 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3426 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5064 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1555 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3394 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/4340 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3373 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3453 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4834 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3850 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3119 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3374 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3394 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/952 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3415 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3652 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -500,7 +500,7 @@ static void logProcessPoolState(const WebProcessPool& pool)
 #if HAVE(VIDEO_RESTRICTED_DECODING)
 #if PLATFORM(MAC)
-    if (!isFullWebBrowser() || isRunningTest(WebCore::applicationBundleIdentifier())) {
+    if (!isFullWebBrowser()) {
```
I think it would be good to remove this check now that isFullWebBrowser() is returning the expected value for Safari on macOS.

szewai replied:
Sorry I don't get it: which check should we remove? `if (!isFullWebBrowser()) {` ? Then we will create `trustdExtensionHandle` unconditionally?

szewai replied:
Per clarified offline that since this is check always returns true, to not break existing behavior, we should just remove it. Updated patch.

pvollan commented on Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm:
```diff
@@ -243,21 +248,37 @@ bool isParentProcessAFullWebBrowser(AuxiliaryProcess& auxiliaryProcess)
 
+#if ENABLE(APP_BOUND_DOMAINS)
     static bool fullWebBrowser = WTF::processHasEntitlement("com.apple.developer.web-browser"_s);
+#elif PLATFORM(MAC)
+    static bool fullWebBrowser;
+    static std::once_flag once;
+    std::call_once(once, [] {
+        NSURL *currentURL = [[NSBundle mainBundle] bundleURL];
+        NSArray<NSURL *> *httpURLs = [[NSWorkspace sharedWorkspace] URLsForApplicationsToOpenURL:[NSURL URLWithString:@"http:"]];
+        bool canOpenHTTP = [httpURLs containsObject:currentURL];
+        NSArray<NSURL *> *httpsURLs = [[NSWorkspace sharedWorkspace] URLsForApplicationsToOpenURL:[NSURL URLWithString:@"https:"]];
+        bool canOpenHTTPS = [httpsURLs containsObject:currentURL];
+        fullWebBrowser = canOpenHTTPS && canOpenHTTP;
+    });
```
Will for example MiniBrowser be classified as a full Web browser with this implementation? Will the return value be different in the UI process and the WebContent process?

szewai replied:
> Will for example MiniBrowser be classified as a full Web browser with this implementation?

No, because MiniBrowser does not register for these two schemes. This is the heuristics used for identifying browsers by system, see https://developer.apple.com/documentation/xcode/preparing-your-app-to-be-the-default-browser-or-email-client, so I think this is a bug in MiniBrowser.

> Will the return value be different in the UI process and the WebContent process?

This function is not supposed to be called in web process and we have assertion above:
`ASSERT(!isInWebKitChildProcess());`
Web process should use `isParentProcessAFullWebBrowser`.

pvollan reviewed (approved):
Looks good! R=me.

szewai committed:
Implement isFullWebBrowser check on macOS
https://bugs.webkit.org/show_bug.cgi?id=255668
rdar://108271078

Reviewed by Per Arne Vollan.

isFullWebBrowser() is available on Cocoa platforms but only returns meaningful value on non-macOS platforms, because it
is implemented by checking an entitlement that is not avaiable on macOS. This patch implements the check on macOS. To
avoid breakage, this patch does not change any behaivor, by ensuring the existing callsites get the same value as before
this patch.

Also, `isFullWebBrowser()` actually returns true when app is browser *or* test app, so this patch renames it to
`isFullWebBrowserOrRunningTest()`.

* Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.h:
* Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm:
(WebKit::determineTrackingPreventionStateInternal):
(WebKit::isFullWebBrowserOrRunningTest):
(WebKit::isFullWebBrowser): Deleted.
* Source/WebKit/UIProcess/API/APIHTTPCookieStore.cpp:
(API::HTTPCookieStore::filterAppBoundCookies):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::loadData):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::setIsNavigatingToAppBoundDomainAndCheckIfPermitted):

Canonical link: https://commits.webkit.org/263198@main

youennf commented on Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm:
```diff
@@ -99,10 +99,15 @@ static bool determineTrackingPreventionStateInternal(bool appWasLinkedOnOrAfter,
         return false;
 
-    if (!isFullWebBrowser(bundleIdentifier) && !hasRequestedCrossWebsiteTrackingPermission())
+    if (!isFullWebBrowser && !hasRequestedCrossWebsiteTrackingPermission())
```
isFullWebBrowser is a bit misleading, it would be more something like isTrackingPreventionEntitledApplication, which is a bit long.
